agent/consul/leader_peering_test.go

```diff
@@ -94,55 +94,61 @@ func testLeader_PeeringSync_Lifecycle_ClientDeletion(t *testing.T, enableTLS boo
 			c.TLSConfig.GRPC.KeyFile = "../../test/hostname/Betty.key"
 		}
 	})
-	testrpc.WaitForLeader(t, s2.RPC, "dc2")
+	testrpc.WaitForLeader(t, dialer.RPC, "dc2")
 
-	// Create a peering at s2 by establishing a peering with s1's token
+	// Create a peering at dialer by establishing a peering with acceptor's token
 	ctx, cancel = context.WithTimeout(context.Background(), 3*time.Second)
 	t.Cleanup(cancel)
 
-	conn, err = grpc.DialContext(ctx, s2.config.RPCAddr.String(),
-		grpc.WithContextDialer(newServerDialer(s2.config.RPCAddr.String())),
+	conn, err = grpc.DialContext(ctx, dialer.config.RPCAddr.String(),
+		grpc.WithContextDialer(newServerDialer(dialer.config.RPCAddr.String())),
 		grpc.WithInsecure(),
 		grpc.WithBlock())
 	require.NoError(t, err)
 	defer conn.Close()
 
-	s2Client := pbpeering.NewPeeringServiceClient(conn)
+	dialerClient := pbpeering.NewPeeringServiceClient(conn)
 
 	establishReq := pbpeering.EstablishRequest{
-		PeerName:     "my-peer-s1",
+		PeerName:     "my-peer-acceptor",
 		PeeringToken: resp.PeeringToken,
 	}
-	_, err = s2Client.Establish(ctx, &establishReq)
+	_, err = dialerClient.Establish(ctx, &establishReq)
 	require.NoError(t, err)
 
-	p, err := s2Client.PeeringRead(ctx, &pbpeering.PeeringReadRequest{Name: "my-peer-s1"})
+	p, err := dialerClient.PeeringRead(ctx, &pbpeering.PeeringReadRequest{Name: "my-peer-acceptor"})
 	require.NoError(t, err)
 
 	retry.Run(t, func(r *retry.R) {
-		status, found := s2.peerStreamServer.StreamStatus(p.Peering.ID)
+		status, found := dialer.peerStreamServer.StreamStatus(p.Peering.ID)
+		require.True(r, found)
+		require.True(r, status.Connected)
+	})
+
+	retry.Run(t, func(r *retry.R) {
+		status, found := acceptor.peerStreamServer.StreamStatus(p.Peering.PeerID)
```

Comment:
This was the change to wait for the stream to be connected before deleting

Reply:
Though there's still a race that can happen outside of tests. The issue is this sequence:
1. `StreamResources` handler is invoked on acceptor, and peering exists, so the connection is allowed
2. Peering is deleted on acceptor
3. HandleStream begins on acceptor

Now the stream is tracked but won't be torn down until the leader routine wakes up due to a change in the list of peerings